[ENH] create `rbf_forecaster.py` using `RBFTransformer` Neural networks

phoeenniixx opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->
Fixes #7271

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
implements `RBFForecaster` using `RBFTransformer` Layers.
#### Does your contribution introduce a new dependency? If yes, which one?
no

phoeenniixx commented:
when i did the `pre-commit` test locally, it itself added the `docs/source/examples` change to it. But here it is failing due to that🫠
I even updated the precommit using ` pre-commit autoupdate --repo https://github.com/pre-commit/pre-commit-hooks` and now the `pre-commit-config.yaml` is also changed...
so should i revert the these change? if yes how? Not sure how to revert just one file...

phoeenniixx commented:
when i did the `pre-commit` test locally, it itself added the `docs/source/examples` change to it. But here it is failing due to that🫠
I even tried to solve this by updating the precommit using ` pre-commit autoupdate --repo https://github.com/pre-commit/pre-commit-hooks` and now the `pre-commit-config.yaml` is also changed...
so should i revert the these change? if yes how? Not sure how to revert just one file...

phoeenniixx commented:
Hi @fkiraly, am i missing something here? I added `torch` to `python_dependencies` still I am getting this error here:
`ModuleNotFoundError: No module named 'torch'`
Please help me out...

phoeenniixx commented:
I think we can add the `api-reference` to `Deep learning based forecasters` sub-class.. wdyt?

fkiraly commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,601 @@
+    gamma : float, optional (default=1.0)
+        Scaling factor controlling the spread of the RBF.
+    rbf_type : {"gaussian", "multiquadric", "inverse_multiquadric"}
+                optional (default="gaussian")
```
this might not display correctly in the docstring, have you checked the readthedocs build? ("details" next to the readthedocs CI element)

phoeenniixx replied:
The doc-build is not showing any error and when i tried the "details" part, I gets redirected to this site: https://sktime--7334.org.readthedocs.build/en/7334/

So to see any errors I need to register there?
I have done that as well... 
maybe after I get registered as a user I will be able to see any issue?

fkiraly commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,601 @@
+class RBFForecaster(BaseForecaster):
+    r"""Forecasting model using RBF transformations and 'NN' layers for time series.
+
+    This forecaster uses an RBF layer to transform input time series data into a
```
this should probably say what exactly the architecture is, the user will not look at the code - so a description of the architecture or a link to a paper would be nice.

fkiraly commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,601 @@
+                optimizer.step()
+                total_loss += loss.item()
+
+        self._is_fitted = True
```
not necessary to do that, the boilerplate layer (`fit`) takes care of state handling.

fkiraly commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,601 @@
+        """
+        import torch
+
+        if not self._is_fitted:
```
same as above, done in the boilerplate

fkiraly reviewed (changes requested):
Interesting contribution!

I would kindly request:

* document the structure of the model in the docstring
* this should be added to one of the deep learning subcategories in the API reference, `foercasting.rst`
* have a look if the `BaseDeepNetworkPyTorch` class by @yarnabrina could be a base class that could help avoid some duplications in the interface layer to `pytorch`, but htis is optional, non-blocking

phoeenniixx commented:
@fkiraly any updates?

phoeenniixx commented:
Hi, @fkiraly any updates?

phoeenniixx commented:
Hi @fkiraly, I think this is it from my end, I think the forecaster is complete, please review it and tell me if you want any changes!
I have changed the whole structure of the file so that `RBFLayer` is now independently available and the user can use just the Layer as well if he wants to create his own Forecaster using the layer and donot need to import the whole forecaster to do that.

phoeenniixx commented:
Hi @fkiraly, any updates?

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+        X : optional
+            Additional exogenous data (not used).
+        """
+        import torch
```
I think this import is unnecessary. Since if the interpreter reaches this line, torch is already imported or?

phoeenniixx replied:
yeah sorry I missed it, added while I was creating and testing the class and might have missed it (as much as I can remember)

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+
+        return pd.Series(predictions, index=fh_abs, name=self._y.name)
+
+    def plot_predictions(self, y_train, y_test, y_pred):
```
Why did you add this plot method? 

I think the `sktime` plotting methods that are not bound to specific estimators are sufficient.

phoeenniixx replied:
thought it might be useful that the forecaster have its own plotting method😅 I will remove it

benHeid reviewed (changes requested):
Thanks for the PR. I am not that familiar with RBF networks. But please check my comments below.

phoeenniixx commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+        x_np = x.detach().cpu().numpy()
+        self.rbf_transformer.fit(x_np)
+
+        x_rbf = self.rbf_transformer.transform(x_np)
```
>How expensive is fitting an rbf_transformer? 

rbf_transformer is fitting is just creating centers. it creates its own centres if not provided using np.linspace.

>Thus, you might move it to the _fit method directly?

So, you want me to actually remove the classes `RBFLayer` etc? I added it as rn the forecaster has a predefined structure and the user may want only the layer to create his own "custom" forecaster... So, should I remove it and make the forecaster only?

phoeenniixx replied:
>stacking those layer is impossible...

I am just using one layer in `RBFNetwork` as I think that much is sufficient and then just adding linear layers. Are you saying that stacking one RBFLayer over another is impossible? So maybe yeah as once the data passes through it, it is transformed by the rbf_transformer and now the data is changed, I dont think we need to use this layer again on that data as it is already transformed and we got what we want.

Am I interpreting what you said correctly?

benHeid replied:
So currently the transformation is done on each batch. I.e. also during predict this transformation is fitted and then applied. I am not familiar enough with the algorithm to know if it is correct that fit and transform is done one each  batch. 
Furthermore,transforming everything before starting training would be more efficient, since you do this right now for each batch multiple times..

phoeenniixx replied:
Ok, I will then just transform it only once before training

phoeenniixx replied:
Hi @benHeid, I had some doubts so I read a little more about RBF NNs and got to know about this:

A radial basis function (RBF) network contains three layers such as an input layer, a hidden layer with a non-linear [RBF](https://www.sciencedirect.com/topics/engineering/radial-basis-function) activation function and a linear output layer. The second layer /hidden layer perform a nonlinear mapping from the input space into a higher dimensional space by using a [Gaussian](https://www.sciencedirect.com/topics/physics-and-astronomy/gaussian-distribution) or some other kernel function. Output layer:-The final layer performs a weighted sum with a linear output.

(source: https://www.sciencedirect.com/topics/engineering/radial-basis-function-network)

I checked in multiple sites, everywhere they show the same structure. This is very similar to mine that I have already implemented. So, what should we do? Do the transformation only once or follow this structure?

The transformation does the same thing explained here: apply the RBF functions on the data

phoeenniixx replied:
lmk what do you think?

benHeid replied:
Mhm.. Can you implement the RBF transformation directly in the layer? So that no transformation of the tensors to numpy vectors are necessary? In that case, the whole network should be differentiable and everything should also be a lot faster.

phoeenniixx replied:
but the `rbf_transformer` wants the input to be in `sktime` compatible format (which is not tensors) and if i want the transformations to be in the  layer directly, that means as much as i can understand, we can do one thing that we first pass the `sktime` compatible data ( the way it was entered in the forecaster) to the transformer and then turn them into `tensors`, then only once the `numpy` to `tensor` transformations are done, but it can be done only once before the loop, after that once it is turned back to tensors, to do the transformation in next epoch, we will have to turn the data back to `numpy` so that it can be fed to the transformer, So, in the end the problem is not solved.

the solution to this is we donot use the `rbf_transformer` rather do the transformations in the function only and not call the transformer.

benHeid replied:
I would prefer to not call the sktime transformer and do the transformations directly in the function. Then it would not be necessary to transform everything to numpy within in the layer or? 

Would this also solve the issue that the optimizer needs to be instantiated in the loop?

phoeenniixx replied:
> I would prefer to not call the sktime transformer and do the transformations directly in the function. Then it would not be necessary to transform everything to numpy within in the layer or?

Yes, in the recent commits I have removed the transformer and doing the transformations directly in the `forward` function of `RBFLayer` and now it works only on the tensors and are not changed to numpy

> Would this also solve the issue that the optimizer needs to be instantiated in the loop?

Yes it does! As now the transformations are done in the Layer only and we don't need the Lazy initialization technique for the dense layers. I have now instantiated outside the loop only

phoeenniixx commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+    criterion : {"mse", "l1", "poisson", "bce", "crossentropy"}
+               optional (default="mse")
+        Loss function to use during training.
+    use_cuda : bool, optional (default=False)
```
ok will do that

phoeenniixx commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+        for epoch in range(self.epochs):
+            total_loss = 0
+            for batch_X, batch_y in dataloader:
+                if optimizer is None:
```
I created the the layers using "lazy initialisation" that i read somewhere, as the linear layer shape is defined after the fitting of `rbf-layer`, so the parameters are not known at the initialisation. That's why I added it in the loop only. If you want me to remove the RBFLayer class then it might change (not sure it will, as then i will remove the lazylinear class as well and everything will be done in _fit or in __init__ only)

benHeid replied:
I mean the initialisation of the optimiser. Is the needed in the for loop?

phoeenniixx replied:
Actually the the parameters required to initialise the optimizer are not present before, as the shapes and all of the layers depends on the output of the transformer and rn it is happening in the loop only, so although optimizer is initialised only once in the first epoch only after that due to if statement it is not executed, it is required in the loop only, but now as you are saying I should transform it only once before fitting, that would change the things ig... I will look into it

phoeenniixx commented:
Thanks for the review @benHeid, just have a doubt:
Should I remove the RBFLayer and LazyLinear classes and do everything in the same parent class only?
(please see the answers to the reviews as well!)

benHeid commented:
> I will make changes after your confirmation

See my comment in the relevant comment. After reading your answer, I think the best is to implement the RBF transformations directly in the layer without using the sktime transformations. This should be a lot faster and also the layer is still differentiable.

phoeenniixx commented:
Hi @benHeid, @fkiraly, the `test-unix-pandas` are failing which i think is unrelated to this PR
The fails are:
```
FAILED sktime/tests/test_softdeps.py::test_est_construct_if_softdep_available[TSFreshRelevantFeatureExtractor] - ImportError: cannot import name 'cwt' from 'scipy.signal' (/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/signal/__init__.py)
FAILED sktime/tests/test_softdeps.py::test_est_construct_if_softdep_available[TSFreshFeatureExtractor] - ImportError: Error in TSFreshFeatureExtractor.get_test_params, return must be valid param dict for class, or list thereof, but attempted construction raised a exception. Problematic parameter set: {'disable_progressbar': True, 'show_warnings': False, 'default_fc_parameters': 'minimal'}. Exception raised: cannot import name 'cwt' from 'scipy.signal' (/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/signal/__init__.py)
FAILED sktime/tests/test_softdeps.py::test_est_get_params_without_modulenotfound[TSFreshRelevantFeatureExtractor] - ImportError: cannot import name 'cwt' from 'scipy.signal' (/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/signal/__init__.py)
```

there seems to be a problem with `TSFreshRelevantFeatureExtractor`.
Please look into it...
Thanks!!

phoeenniixx commented:
@benHeid, please review the PR, I have removed the transformer and added the logic for RBF function in the classes only so now it doesnt need to convert numpy to tensors and back repeatedly.

fkiraly commented:
Ouch, looks like a bit of a `rebase` accident. I would recommend to avoid `rebase` altogether for contributions to `sktime` (or other open source) - we squash pull requests anyway, so history within your branch is not important.

To recover, I would suggest you go through the 10 files that show up as changed here
https://github.com/sktime/sktime/pull/7334/files
and are not actually within the scope of your branch, and copy the versions from `main` over them, then commit.

That might be the quickest, least painful way to recover from this.

Alternatively, you could:

* copy the changed files into a safe location, check out `main`, copy that over your branch, then copy the file over your branch, then commit, but this is likely more time
* use `git` magic to revert the rebase or selectively pick changes from the file, but this has more cognitive complexity if you are not a `git` pro and will also likely take more time

phoeenniixx commented:
Thanks @fkiraly, I tried the first way, is this ok?

phoeenniixx commented:
Also pinging @benHeid for the review :)

benHeid commented:
> I am not sure it worked, only one file is removed now: `sktime/utils/_testing/forecasting.py`, rest of them are still here T_T

You can do this also manually and copying the files from main that you do not want to change. I think this is what @fkiraly proposed in his comment.

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+                f"rbf_type must be one of {valid_rbf_types}, got {self.rbf_type}"
+            )
+
+    def forward(self, x):
```
Perhaps you can reduce the code duplication in RBFTransformer by importing this Layer in case torch should be used by RBFTransformer. (Might be a separate PR).

benHeid replied:
Would appreciate an Issue, after the pr is merged.

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+        layers = []
+        prev_size = hidden_size
+
+        for size in linear_layers:
```
Can you rename it to `hidden_layers`? `linear_layers` might be confusing since you are always adding a ReLU after the layer.

benHeid replied:
Ok after reading the documentation this might be okay also as it is. However, still slightly prefer an other name. What do you think?

phoeenniixx replied:
well `hidden_layers` would be better in my opinion as well as I think it is more widely, I will change it.

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+        prev_size = hidden_size
+
+        for size in linear_layers:
+            layers.extend([nn.Linear(prev_size, size), nn.ReLU(), nn.Dropout(0.1)])
```
You might make the activation and the dropout configurable via hyperparams. So the user can easily exchange ReLU or dropout

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+       - Output shape: (batch_size, hidden_size)
+    3. Sequential layers:
+       - One or more fully connected linear layers
+       - Each linear layer uses the LazyLinear module, which initializes the weights
```
I think, we removed the Lazy or?

phoeenniixx replied:
yes sorry, i forgot to remove it from the docstring, will do this right away

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+       - ReLU activation is applied after each linear layer
+       - Dropout with a rate of 0.1 is applied after each linear layer
+       - The final linear layer produces the output prediction
+       - Output shape: (batch_size, 1)
```
Should the output be dependent on the forecasting_horizon?

benHeid replied:
After seeing the predict method, I understand that this models works autoregressive.. Need it to be operated in that way?

phoeenniixx replied:
So yes this is an autoregressive implementation of RBF network (aldo called RBF-AR) as it uses the RBF kernels as well as the autoregression. If you want we can change it?

phoeenniixx commented:
> You can do this also manually and copying the files from main that you do not want to change. I think this is what @fkiraly proposed in his comment.

Actually i tried the same, and it worked for one file atleast (previously there were 10 changed files, now there are 9) but rest are still  showing as edited,

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+            Initialized device object
+        """
+        if isinstance(device, str):
+            if device not in ["cpu", "cuda"]:
```
can you add "mps" here?

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,575 @@
+
+        self.use_cuda = use_cuda
+
+        if self._check_torch_availability():
```
I don't think you need this. Since you already specified torch as dependency.

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+                device = "cpu"
+        return torch.device(device)
+
+    def _check_torch_availability(self):
```
You probably don't need this method.

benHeid replied:
See comment above where I also pinged fkiraly

benHeid commented on sktime/forecasting/rbf_forecaster.py:
```diff
@@ -0,0 +1,523 @@
+        """Check if torch is available and raise appropriate error if not."""
+        if not _check_soft_dependencies("torch", severity="none"):
+            raise ModuleNotFoundError(
+                "torch is a required optional dependency for RBFForecaster. "
```
I think it is not optional

benHeid commented:
> Thank you @benHeid for the review! I have added "modes" to the `RBFForecaster` that now allows the users to choose between the "direct" and "ar" mode, here ar is the autoregressive mode, (see the comments above) as I thought it would be better to have user make the choice of the mode. Please check and tell me if it would be ok to add this feature? I have also made other changes you asked for

Nice. Can you test this option, by adding this option to one of the two test sets in get_test_params?

phoeenniixx commented:
> > Thank you @benHeid for the review! I have added "modes" to the `RBFForecaster` that now allows the users to choose between the "direct" and "ar" mode, here ar is the autoregressive mode, (see the comments above) as I thought it would be better to have user make the choice of the mode. Please check and tell me if it would be ok to add this feature? I have also made other changes you asked for
> 
> Nice. Can you test this option, by adding this option to one of the two test sets in get_test_params?

I think I did that..? (please check line 611)
The new arg is "mode"

phoeenniixx commented:
Hi @benHeid, I think I have made the changes you requested, please review them

benHeid commented:
@fkiraly are your requested changes addressed?

phoeenniixx commented:
I think the problem is internal, maybe the runtime didn't load properly that's why that test has failed?
```
FAILED sktime/detection/tests/test_plotting.py::test_plot_time_series_with_profiles - _tkinter.TclError: Can't find a usable init.tcl in the following directories: 
    {C:\hostedtoolcache\windows\Python\3.10.11\x64\tcl\tcl8.6}

C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tcl8.6/init.tcl: couldn't read file "C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tcl8.6/init.tcl": No error
couldn't read file "C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tcl8.6/init.tcl": No error
    while executing
"source C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tcl8.6/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"


This probably means that Tcl wasn't installed properly.
```
